### ISSUE TITLE
Regression with unification within fn (?)

### ISSUE THREAD

**Issue by flodiebold** [rust-lang/chalk#658]
See https://github.com/rust-analyzer/rust-analyzer/issues/6628.

I've got this down to the following Chalk test:
```rust
#[test]
fn ambiguous_unification_in_fn() {
    test! {
        program {
            trait FnOnce<Args> {
                type Output;
            }

            struct MyClosure<T> {}
            impl<T> FnOnce<(T,)> for MyClosure<fn(T) -> ()> {
                type Output = ();
            }
        }
        goal {
            exists<int T, U> {
                MyClosure<fn(&'static U) -> ()>: FnOnce<(&'static T,)>
            }
        } yields {
            "Unique"
        }
    }
}
```

This succeeded before #609 and now fails with an ambiguous result.

CC @jackh726  @daboross

**Comment by jackh726** [comment]
A more minimal example:
```
#[test]
fn ambiguous_unification_in_fn() {
    test! {
        program {
            trait FnOnce<Args> {}

            struct MyClosure<T> {}
            impl<T> FnOnce<T> for MyClosure<fn(T) -> ()> {}
        }
        goal {
            exists<T, U> {
                MyClosure<fn(U) -> ()>: FnOnce<T>
            }
        } yields {
            "Unique"
        }
    }
}
```

**Comment by jackh726** [comment]
However, this passes:
```
#[test]
fn ambiguous_unification_in_fn() {
    test! {
        program {
            trait FnOnce<Args> {}

            struct MyClosure<T> {}
            impl<T> FnOnce<T> for MyClosure<fn(T) -> ()> {}
        }
        goal {
            forall<V> {
                exists<U> {
                    MyClosure<fn(U) -> ()>: FnOnce<V>
                }
            }
        } yields {
            "Unique"
        }
    }
}
```

**Comment by jackh726** [comment]
For reference, using d112f2e69ded929dd, the complete output of the test is: `Unique; for<?U0> { substitution [?0 := ^0.0, ?1 := ^0.0], lifetime constraints [] }`

(The output of the non-regressed test is `Unique; substitution [?0 := !1_0], lifetime constraints []`)

**Comment by jackh726** [comment]
Alright, so, here's the problem:
To solve the goal, we have a clause `for<type> Implemented(MyClosure<"rust" for<0> [?0 := ^1.0, ?1 := 0]>: FnOnce<^0.0>)` that we want to use for our goal `Implemented(MyClosure<"rust" for<0> [?0 := ?0, ?1 := 0]>: FnOnce<?1>)`.

First, we instantiate binders existentially: `Implemented(MyClosure<"rust" for<0> [?0 := ?2, ?1 := 0]>: FnOnce<?2>)`.

Then, we try to relate that with our goal with the `Invariant` variance. Without going into all the details, but because of [this](https://github.com/rust-lang/chalk/blob/2de2e4ad0da28ff2d700bc283b9ff34a581dabf0/chalk-solve/src/infer/unify.rs#L393), we end up with a strand with two goals: `(?0 <: ?2)` and `(?2 <: ?0)`.

To solve either constraint, we have to solve `(^0.0 <: ^0.1)`, which flounders.


Now, what's the best solution here? I'm not exactly sure. A naive solution would be to make [this](https://github.com/rust-lang/chalk/blob/2de2e4ad0da28ff2d700bc283b9ff34a581dabf0/chalk-solve/src/infer/unify.rs#L396) and the one below it to use `Invariant` when the ambient variance is `Invariant`. However, importantly, that breaks [https://github.com/rust-lang/chalk/blob/2de2e4ad0da28ff2d700bc283b9ff34a581dabf0/tests/test/subtype.rs#L75](this test) (it reverts back to the previous behavior of `!1_0: !1_1` and `!1_1: !1_0`).

Unfortunately, I don't know off the top of my head what to do here. I'll have to look more into it, but some input from @matthewjasper and/or @nikomatsakis would be helpful here.

**Comment by jackh726** [comment]
Okay, so I did find one solution that *works*, but seems hacky to me. Basically, when relating two things, if we find that we *would* be creating something like `?0 <: ?1` and `?1 <: ?0`, we instead just unify them. To test this, I just modified [`push_subtype_goal`](https://github.com/rust-lang/chalk/blob/2de2e4ad0da28ff2d700bc283b9ff34a581dabf0/chalk-solve/src/infer/unify.rs#L1073) to check the list of goals for `Subtype(b, a)`; if found, remove it and `relate_ty_ty(Variance::Invariant, a, b)`.

It's perhaps important to note that rustc [always](https://github.com/rust-lang/rust/blob/774bce7f5e9d0cc99a30cf37be5d175a3b530beb/compiler/rustc_infer/src/infer/nll_relate/mod.rs#L311) relates two  inference variables as `Invariant`.
